Porting UPP for gfs v16  on wcoss2

WenMeng-NOAA opened the pull request:
The changes are added for porting UPP for gfs v16 on wcoss2.

Commit:
Add changes for building executable on wcoss2.

Commit:
Add changes for buding upp lib on wcoss2.

Commit:
Add changes for building upp lib on wcoss2.

Commit:
Update hdf5 and netcdf modules.

Commit:
Upgrade manage_externals to new version which works for python 3.8.5.

Commit:
Remove jlogfile in workflow scripts.

Commit:
Changes for overriding module versions from upstream setting.

Commit:
Changes for overriding module versions from upstream setting for
building upp lib.

WenMeng-NOAA committed:
Update for POSTGRB2TBL default setting.

WenMeng-NOAA committed:
Update for LMod compliance.

WenMeng-NOAA committed:
Remove v8.0.0-wcoss2 and lib-wcoss2; correct typo in upp_wcoss2.lua and
post_wcoss2.lua.

aerorahul commented:
@WenMeng-NOAA why are manage_externals changes required here?

WenMeng-NOAA commented:
@aerorahul @KateFriedman-NOAA found a new version of manage_externals is needed for the global-workflow checking gtg code.

KateFriedman-NOAA commented:
> @aerorahul @KateFriedman-NOAA found a new version of manage_externals is needed for the global-workflow checking gtg code.

@WenMeng-NOAA Can the GTG checkout happen without manage_externals? I use whatever method you provide on the workflow side to do the checkouts. We're trying to move away from manage_externals now for a few reasons. Can we clone the GTG with submodule commands instead? Thanks!

KateFriedman-NOAA commented on jobs/J_NCEPPOST:
```diff
@@ -27,7 +27,7 @@ export KEEPDATA=${KEEPDATA:-NO}
 # File To Log Msgs
 ####################################
-export jlogfile=${jlogfile:-${COMROOT}/logs/jlogfiles/jlogfile.${job}.${pid}}
+#export jlogfile=${jlogfile:-${COMROOT}/logs/jlogfiles/jlogfile.${job}.${pid}}
```
Perhaps just remove this line now that we're retiring the use of jlogfiles?

WenMeng-NOAA replied:
@KateFriedman-NOAA done!

KateFriedman-NOAA reviewed (changes requested):
Non-manage-externals-related changes look good and cover the changes Wen and I discussed when I initially tested the post jobs on WCOSS2. I am rebuilding my copy of the UPP with the recent LUA updates and will do the following for this PR:

1. retest post jobs on Dogwood with updated UPP copy and have Wen look at output
2. update my copy of the UPP to use submodules instead of manage_externals; will provide needed updates for Wen to incorporate

Stay tuned...

WenMeng-NOAA commented:
> > @aerorahul @KateFriedman-NOAA found a new version of manage_externals is needed for the global-workflow checking gtg code.
> 
> @WenMeng-NOAA Can the GTG checkout happen without manage_externals? I use whatever method you provide on the workflow side to do the checkouts. We're trying to move away from manage_externals now for a few reasons. Can we clone the GTG with submodule commands instead? Thanks!

@KateFriedman-NOAA Yes, @aerorahul has been helping me on configurating gtg code checking out as a submodule.

KateFriedman-NOAA commented:
@WenMeng-NOAA Please see the following log from rebuilding the UPP after recent updates, please confirm the build went as intended:

/lfs/h2/emc/global/noscrub/Kate.Friedman/git/feature-ops-wcoss2/sorc/logs/build_ncep_post_test5.log

Thanks!

Note: this is before I make any changes for the GTG checkout

WenMeng-NOAA committed:
Remove jflofile per Kate's comments

WenMeng-NOAA commented:
> @WenMeng-NOAA Please see the following log from rebuilding the UPP after recent updates, please confirm the build went as intended:
> 
> /lfs/h2/emc/global/noscrub/Kate.Friedman/git/feature-ops-wcoss2/sorc/logs/build_ncep_post_test5.log
> 
> Thanks!
> 
> Note: this is before I make any changes for the GTG checkout

@KateFriedman-NOAA Your build log look good to me. Thanks!

KateFriedman-NOAA commented:
> @KateFriedman-NOAA Your build log look good to me. Thanks!

Thanks for confirming @WenMeng-NOAA ! I'm working on the patch to retire manage_externals and use submodules instead for the GTG checkout. Will provide the patch to include in this PR when ready.

KateFriedman-NOAA commented:
@WenMeng-NOAA A quick update... @aerorahul and I are still working on the patch (ironing out a detail) but I have removed manage_externals in my fork copy of the `post_gfsv16_wcoss2` branch and added the initial submodule setup. See my changes thus far at this link:

https://github.com/WenMeng-NOAA/UPP/compare/post_gfsv16_wcoss2...KateFriedman-NOAA:post_gfsv16_wcoss2

I hacked things on the workflow side a tad to get the right tag version of the GTG checked out (the detail to iron out) and tested the build with GTG changes in place. Please see the following log and confirm that my build with GTG worked:

/lfs/h2/emc/global/noscrub/Kate.Friedman/git/feature-ops-wcoss2/sorc/logs/build_ncep_post_test6_gtg.log

I will submit a PR into your fork `post_gfsv16_wcoss2` branch with the patch when ready. Thanks!

WenMeng-NOAA commented:
> /lfs/h2/emc/global/noscrub/Kate.Friedman/git/feature-ops-wcoss2/sorc/logs/build_ncep_post_test6_gtg.log

Your build log looks good to me. Thanks!

KateFriedman-NOAA commented:
> Your build log looks good to me. Thanks!

Thanks @WenMeng-NOAA ! I just reran all of my gdaspost* jobs and the gfspost023 job in my Dogwood test. Please see the following logs and output and let me know if it looks ok:

gdaspost*:
/lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/logs/2021082400/gdaspost0*log
/lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/gdas.20210824/00/atmos

gfspost023 (just redid the f213 files):
/lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/logs/2021082400/gfspost023.log
/lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/gfs.20210824/00/atmos/*f213*

FYI, I will be running a cleaner fresh test next week for everyone to examine, so I'll send logs/output to look at again when I do that test. Thanks!

WenMeng-NOAA commented:
> > Your build log looks good to me. Thanks!
> 
> Thanks @WenMeng-NOAA ! I just reran all of my gdaspost* jobs and the gfspost023 job in my Dogwood test. Please see the following logs and output and let me know if it looks ok:
> 
> gdaspost*: /lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/logs/2021082400/gdaspost0*log /lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/gdas.20210824/00/atmos
> 
> gfspost023 (just redid the f213 files): /lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/logs/2021082400/gfspost023.log /lfs/h2/emc/ptmp/Kate.Friedman/comrot/testcyc768b/gfs.20210824/00/atmos/_f213_
> 
> FYI, I will be running a cleaner fresh test next week for everyone to examine, so I'll send logs/output to look at again when I do that test. Thanks!

@KateFriedman-NOAA The runtime logs and UPP datasets look good to me. Thanks!

KateFriedman-NOAA commented:
@WenMeng-NOAA I submitted [PR #5](https://github.com/WenMeng-NOAA/UPP/pull/5) from my fork to your fork with updates for converting the GTG checkout from manage_externals to git submodules. I wasn't able to add reviewers or assignees or labels.

WenMeng-NOAA commented:
@KateFriedman-NOAA Your PR looks good to me. Would you like to have someone else review it?

KateFriedman-NOAA commented:
> @KateFriedman-NOAA Your PR looks good to me. Would you like to have someone else review it?

Nope, since it works for you and both @aerorahul and myself tested it, it should be good to submit my PR into your branch. Thanks!

KateFriedman-NOAA committed:
Change GTG checkout to submodule and remove manage_externals (#5)

* Retire manage_externals for GTG checkout

- remove manage_externals copy from repo
- remove associated Externals.cfg file
- replacing GTG checkout with submodules

Refs: #363

* Add UPP_GTG submodule

- created .gitmodules file via git submodule add command:
git submodule add https://github.com/NCAR/UPP_GTG sorc/post_gtg.fd
- new post_gtg.fd subfolder added (empty before submodule checkout)

Refs: #363

* Updating submodule pointer of GTG to ncep_post_gtg.v2.0.4

WenMeng-NOAA committed:
Teak makefile_lib.

WenMeng-NOAA commented:
@KateFriedman-NOAA and @aerorahul It looks the branch is working for inline and offline post in Kate's GFS V16 end-to-end testing on WCOSS2. Do you have more comments  on it? Thanks!

KateFriedman-NOAA reviewed (approved):
Looks good and post jobs in GFS tests are working as intended. Thanks!

aerorahul reviewed (approved):
Defer to @KateFriedman-NOAA. No further comments from me.

KateFriedman-NOAA commented:
@WenMeng-NOAA If you could look at the post output in my free-forecast test that ran overnight and then, assuming it looks good, I'm good for this to be submitted. See this g-w issue comment with output details: https://github.com/NOAA-EMC/global-workflow/issues/399#issuecomment-989998164 . Thanks!
